### PR TITLE
[dLLM] Make FDFO a framework capability for all dLLM algorithms

### PR DESCRIPTION
**Pull request by Hayden727**
## Motivation

This PR supersedes #25280. It generalizes **FDFO (First-Done-First-Out)** scheduling for diffusion LLM (dLLM) inference from a single bespoke algorithm into a **framework-level execution mode that any dLLM algorithm can opt into**, including algorithms that carry cross-step state.

### Problem

In synchronous dLLM batching, all requests in a batch advance in lockstep. High-confidence requests that converge in few denoise steps are forced to wait for low-confidence, long-tail requests before they can leave the batch. This head-of-line blocking leaves the GPU idle and caps throughput. FDFO frees a request the moment its decode block is fully resolved, without waiting for slower requests in the same batch.

### Why a framework capability

#25280 bundled FDFO into a dedicated `LowConfidenceFDFO` algorithm class, which conflated two orthogonal concerns and duplicated the token-selection logic:

- **token-selection strategy** — which masked positions to unmask this denoise step (low-confidence threshold, joint-threshold, …);
- **execution mode** — denoise a whole block synchronously vs. yield to the scheduler after each step (FDFO).

This PR separates them so FDFO works for every algorithm, and adds the per-request state plumbing needed for **stateful** algorithms.

## Design

### Strategy vs. execution split

`DllmAlgorithm` now defines a single-step strategy; the base class owns both execution loops:

- `step(forward_batch, full_logits, states) -> done_flags` — one denoise step; advances `input_ids`/`states` in place and returns, per block, whether it was already complete on entry (so this forward persisted its final KV cache).
- `init_step_state(forward_batch)` / `max_steps(block_size)` — per-block cross-step state and the step budget.
- `run(...)` (base) branches on the mode: synchronous loops `forward → step` until the block resolves; FDFO does one `forward → step` and returns control to the scheduler.

`LowConfidence` (stateless) and `JointThreshold` (stateful) are reduced to `step()`; their internal loops are removed. `LowConfidenceFDFO` is deleted.

### Orthogonal selection

A new `--dllm-fdfo` flag drives `DllmConfig.first_done_first_out_mode`, independent of `--dllm-algorithm`. Any algorithm runs in either mode.

### Cross-step state carrier (stateful algorithms)

- `Req.dllm_incomplete_ids` — the in-progress block tokens (universal).
- `Req.dllm_algo_state` — opaque algorithm-private cross-step state. Stateless algorithms keep it `None`; `JointThreshold` stores `{post_edit_steps, prompt_mask}`. The framework carries it across FDFO rounds via `GenerationBatchResult.dllm_algo_state` and clears it on block completion.

`process_batch_result_dllm_fdfo` manages the per-round lifecycle and releases the masked KV slots of carried blocks.

## Correctness

At **batch size 1**, FDFO and synchronous execution run identical forward shapes, so a correct implementation must produce **byte-identical** output. Verified on `inclusionAI/LLaDA2.0-mini` (H200), temperature 0:

- `LowConfidence`: FDFO == sync, **identical** on all probe prompts (multi-block).
- `JointThreshold` (stateful): FDFO == sync, **identical** on all probe prompts — confirming the cross-step state carry is correct.

At batch > 1, scores fluctuate by ±0.01 between runs for **both** sync and FDFO (kernel batch non-invariance, incl. MoE routing); the two distributions overlap, so this is run-to-run noise rather than an FDFO regression.

## Performance (H200, TP=1, LLaDA2.0-mini, GSM8K, 200 examples)

LowConfidence, output throughput:

| max-running-requests | sync | FDFO | speedup |
|---|---|---|---|
| 4  | 525 tok/s  | 680 tok/s  | **1.30×** |
| 16 | 918–959 tok/s | 1276–1378 tok/s | **~1.45×** |

Scores stay in the same band across modes (sync 0.905–0.92, FDFO 0.90–0.915).

## Checklist

- [x] Format code with pre-commit (black/isort).
- [x] Add unit tests (`test/registered/dllm/test_dllm_batching_fdfo.py`).
- [x] Provide accuracy and speed benchmark results.
- [x] Follow the SGLang code style guidance.











































































































































































































<!-- pr-states:start -->
---
### CI States

Latest PR Test (Base): <!-- slot:pr-test:start -->:x: [Run #29102381559](https://github.com/sgl-project/sglang/actions/runs/29102381559)<!-- slot:pr-test:end -->
Latest PR Test (Extra): <!-- slot:pr-test-extra:start -->:x: [Run #29102381592](https://github.com/sgl-project/sglang/actions/runs/29102381592)<!-- slot:pr-test-extra:end -->
<!-- pr-states:end -->

### CODE REVIEW

File: python/sglang/srt/dllm/mixin/scheduler.py [hunk 1]
```diff
@@ -103,6 +101,73 @@ def process_batch_result_dllm(
...
+    def process_batch_result_dllm_fdfo(
+        self: Scheduler,
+        batch: ScheduleBatch,
+        result: GenerationBatchResult,
+    ):
+        if result.copy_done is not None:
+            result.copy_done.synchronize()
+
+        if result.accept_length_per_req_cpu is None:
+            raise RuntimeError("FDFO dLLM result is missing accept lengths.")
```
Comment (btw616) [inline python/sglang/srt/dllm/mixin/scheduler.py]: It would be better to use `assert` here.

File: python/sglang/srt/managers/scheduler.py [hunk 2]
```diff
@@ -2426,7 +2426,12 @@ def get_next_batch_to_run(self) -> Optional[ScheduleBatch]:
         if self.dllm_config is not None and self.dllm_manager.any_staging_reqs():
             chunked_req_to_exclude.update(self.dllm_manager.staging_queue)
             for req in self.dllm_manager.staging_queue:
-                self.stash_chunked_request(req)
+                if self.dllm_config.first_done_first_out_mode:
+                    if not req.dllm_incomplete_ids:
+                        self.stash_chunked_request(req)
+                    self.req_to_token_pool.free(req)
```
Comment (ClawSeven) [inline python/sglang/srt/managers/scheduler.py:2630]: I think we could reuse the tokens' HBM instead of freeing them and reallocating them in each round, which would reduce allocation overhead.

Anyway, this PR is clean, and the issue can be addressed in a separate PR.

File: python/sglang/srt/dllm/mixin/scheduler.py [hunk 3]
```diff
@@ -103,6 +103,74 @@ def process_batch_result_dllm(
...
+        result: GenerationBatchResult,
+    ):
+        if result.copy_done is not None:
+            result.copy_done.synchronize()
+
+        assert (
+            result.accept_length_per_req_cpu is not None
+        ), "FDFO dLLM result is missing accept lengths."
+
+        self.token_to_kv_pool_allocator.free_group_begin()
```
Comment (ClawSeven) [inline python/sglang/srt/dllm/mixin/scheduler.py]: I think the free logic here is a bit messy. It could be encapsulated in a local function with a clear, descriptive name.

BTW, could you merge process_batch_result_dllm_fdfo and process_batch_result_dllm into a single function?

**Comment by ClawSeven** [comment]
This PR looks great, particularly the step design. Overall, it is pretty clean.

**Comment by ClawSeven** [comment]
FDFO should be the default setting, and we should use the FDFO unit test to replace the previous dLLM unit test.

**Comment by Hayden727** [comment]
Thanks @ClawSeven for the thorough review — really appreciate it! I've addressed all of your suggestions:

### 1. Merge the two result handlers + clean up the free logic
- `process_batch_result_dllm_fdfo` is now merged back into `process_batch_result_dllm`; the sync and FDFO branches share a single handler, and the dispatch in `scheduler.py` is a single call.
- The messy unresolved-block free logic is no longer done in the result handler at all (see below).

### 2. Reuse the tokens' HBM instead of free/realloc each round
Implemented. For an **unresolved (continuing)** FDFO block we now **retain** the `req_pool_idx` and the block's KV slots and re-denoise in place, instead of freeing + reallocating every round:
- `get_next_batch_to_run` only frees the req slot when the block actually resolves.
- The result handler no longer frees the unresolved block's KV.
- `alloc_for_extend` detects such requests (retained `req_pool_idx` + stashed `dllm_incomplete_ids`) and reuses the slots already mapped in `req_to_token`, only drawing fresh slots from the allocator for the rest. Non-dLLM batches take the original allocate-everything path byte-for-byte. It's paging-safe because the dLLM scheduler forces `page_size` to a multiple of `block_size`, so each block is page-aligned and was validly paged when first created.

### Rebase fix found during validation
While validating on GPU I hit a rebase-drift bug: main's #26637/#26659 split `Req.fill_ids` into `full_untruncated_fill_ids` + `fill_len` (the **truncated/committed** length). The FDFO resolved-block write-back must index relative to `fill_len`, because `add_dllm_staging_req` truncates the block to the KV budget under batching (so `fill_len < len(full_untruncated_fill_ids)`). Using the full length wrote resolved tokens to the wrong offset — correct at batch size 1 but garbage at batch size 4. Fixed to match the sync branch.

### Validation (H200, LLaDA2.0-mini, flashinfer, page_size=32)
| Config | Result |
|---|---|
| Sync (`--no-dllm-fdfo`) bs=4 gsm8k | **0.91** |
| FDFO (default, HBM reuse) bs=4 gsm8k | **0.92** |
| FDFO vs sync bs=1 parity (`test_dllm_fdfo_joint_threshold`) | **byte-identical** |

No OOM / KV-pool leak over the full 200-example run.

### Test coverage for the reuse lifecycle
Per your note about asserting the reuse directly, I added a CPU unit test (`test/registered/unit/mem_cache/test_dllm_fdfo_kv_reuse.py`) that checks an unresolved FDFO block reuses its mapped `req_to_token` slots (allocator queried only for fresh requests' tokens), and that an all-reuse batch allocates nothing.

Also: FDFO is now the default (`--dllm-fdfo` → `BooleanOptionalAction`, `--no-dllm-fdfo` to opt out), and the previous dLLM unit test was consolidated into the FDFO batching test. Lint (black/isort) is green now too. Thanks again!

**Comment by ClawSeven** [comment]
@Hayden727 I mean, the token reuse implementation should be in a separate PR. Putting everything in this one will just make it more complicated and harder to merge.

**Comment by Hayden727** [comment]
> @Hayden727 I mean, the token reuse implementation should be in a separate PR. Putting everything in this one will just make it more complicated and harder to merge.

Get, I'll open a separate PR for implement token reuse~

**Comment by Hayden727** [comment]
@ClawSeven I've moved the HBM token-reuse optimization out of this PR into a separate one: #27877. This PR now keeps only the framework changes (merged handler, FDFO as default, test consolidation) plus the `fill_ids` → `full_untruncated_fill_ids`/`fill_len` rebase fix needed for the FDFO path on current main. Thanks!

File: test/registered/dllm/test_dllm_batching_fdfo.py
Comment (btw616) [inline test/registered/dllm/test_dllm_batching_fdfo.py]: It would be better to increase the threshold accordingly.

File: test/registered/dllm/test_dllm_fdfo_joint_threshold.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+]
+
+
+class TestBatchingFDFOJointThreshold(CustomTestCase):
+    """At a single in-flight request, FDFO and synchronous execution run identical
+    forward shapes, so a correct stateful (``dllm_algo_state``) carry must produce
+    byte-identical multi-block output.
+    """
+
+    model = "inclusionAI/LLaDA2.0-mini"
```
Comment (btw616) [inline test/registered/dllm/test_dllm_fdfo_joint_threshold.py]: LLaDA2.0-mini doesn't support JointThreshold. We need to use LLaDA2.1-mini instead.

File: python/sglang/srt/dllm/mixin/scheduler.py [hunk 5]
```diff
@@ -69,24 +69,87 @@ def process_batch_result_dllm(
...
+        def free_unresolved_block_kv(req: Req):
+            # Release the KV slots of the still-masked block so the next FDFO
+            # round can re-denoise it without leaking the previous allocation.
+            old_prefix_len = len(req.prefix_indices)
+            new_fill_len = req.fill_len
+            if new_fill_len > old_prefix_len:
+                kv_indices_to_free = self.req_to_token_pool.req_to_token[
+                    req.req_pool_idx, old_prefix_len:new_fill_len
+                ]
+                self.token_to_kv_pool_allocator.free(kv_indices_to_free)
```
Comment (btw616) [inline python/sglang/srt/dllm/mixin/scheduler.py]: Defining a helper here looks a bit odd. It looks better to just open code it.

File: python/sglang/srt/dllm/mixin/scheduler.py [hunk 6]
```diff
@@ -69,24 +69,87 @@ def process_batch_result_dllm(
         if result.copy_done is not None:
             result.copy_done.synchronize()
 
-        if result.next_token_ids:
-            self.token_to_kv_pool_allocator.free_group_begin()
+        fdfo_mode = self.dllm_config.first_done_first_out_mode
+        if fdfo_mode:
```
Comment (btw616) [inline python/sglang/srt/dllm/mixin/scheduler.py]: It looks better to move the check into the assert.

File: test/registered/dllm/test_dllm_batching_fdfo.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 from sglang.test.ci.ci_register import register_amd_ci, register_cuda_ci
 
-register_cuda_ci(est_time=139, stage="base-b", runner_config="1-gpu-large")
+register_cuda_ci(est_time=200, stage="base-b", runner_config="1-gpu-large")
```
Comment (btw616) [inline test/registered/dllm/test_dllm_batching_fdfo.py]: Why do we need this?

File: test/registered/dllm/test_dllm_batching_fdfo.py
Comment (Hayden727) [reply test/registered/dllm/test_dllm_batching_fdfo.py]: Kept at 350 for now — the measured throughput (~572) is runner-dependent and I'm not certain which machine CI lands on, so raising it risks flakiness on a slower box. Will bump once the runner type is pinned down.

File: docs_new/docs/supported-models/diffusion_language_models.mdx [hunk 8]
```diff
@@ -16,6 +16,19 @@ python3 -m sglang.launch_server \
   --port 30000
 ```
 
+## First-Done-First-Out (FDFO) Scheduling
+
+By default the requests in a batch advance in lockstep, so fast-converging requests must wait for slow long-tail requests before leaving the batch (head-of-line blocking). Add `--dllm-fdfo` to free each request as soon as its block is resolved, which improves throughput. The flag is orthogonal to `--dllm-algorithm` and works with any dLLM algorithm:
```
Comment (btw616) [inline docs_new/docs/supported-models/diffusion_language_models.mdx]: FDFO is now enabled by default. The doc could be updated accordingly.

File: docs_new/docs/advanced_features/server_arguments.mdx [hunk 9]
```diff
@@ -1887,6 +1887,12 @@ Please consult the documentation below and [server_args.py](https://github.com/s
       <td style={{padding: "9px 12px", backgroundColor: "rgba(255,255,255,0.02)"}}>`None`</td>
       <td style={{padding: "9px 12px", backgroundColor: "rgba(255,255,255,0.05)"}}>Type: str</td>
     </tr>
+    <tr>
+      <td style={{padding: "9px 12px", fontWeight: 500, backgroundColor: "rgba(255,255,255,0.02)"}}>`--dllm-fdfo`</td>
+      <td style={{padding: "9px 12px", backgroundColor: "rgba(255,255,255,0.05)"}}>Enable First-Done-First-Out (FDFO) scheduling so completed requests leave the batch immediately instead of waiting for slower requests, eliminating head-of-line blocking. Works with any dLLM algorithm.</td>
+      <td style={{padding: "9px 12px", backgroundColor: "rgba(255,255,255,0.02)"}}>`False`</td>
```
Comment (btw616) [inline docs_new/docs/advanced_features/server_arguments.mdx]: Ditto.

File: test/registered/dllm/test_dllm_batching_fdfo.py
Comment (Hayden727) [reply test/registered/dllm/test_dllm_batching_fdfo.py]: Raised to 450 (original 350 + 100). Matches your ~570 tps on H20 and the ~572 tps measured in CI, so it keeps a safe margin. Thanks!

**Comment by btw616** [comment]
Thanks for the PR! Overall, LGTM.

**Comment by Hayden727** [comment]
> Thanks for the PR! Overall, LGTM.

Thanks for review~

**Comment by Hayden727** [comment]
**Validation** (H100, TP1) — ran the batched accuracy check, since bs=1 parity alone can't catch truncation regressions:

- `inclusionAI/LLaDA2.0-mini`, `--dllm-algorithm LowConfidence --dllm-fdfo`, `--max-running-requests 4`, flashinfer, page_size=32
- `test_dllm_batching_fdfo.py` → **gsm8k score = 0.915** (bs=4), output throughput = 605 tok/s, bs=1 speed = 313 tok/s — both tests pass.

Score is in line with prior runs (~0.91–0.925); no accuracy regression under batched truncation.

**Review by BBuf — APPROVED** [review]
